### PR TITLE
feat(nexus): finish artifact saver

### PR DESCRIPTION
**Pull request by szymon-piechowicz-wandb**
Fixes WB-15188

# Description

Nexus docs: https://weightsandbiases.slack.com/archives/C04MNBGJDBN/p1692678447664909
Upload algorithm docs: https://weightsandbiases.slack.com/archives/C04MNBGJDBN/p1693352398534739
Benchmark: https://weightsandbiases.slack.com/archives/C04MNBGJDBN/p1695764535202659

TODO:
- delete staged files
- write to cache
- S3 multipart upload

# Test plan

TBD

### CODE REVIEW

**Review by raubitsj — APPROVED** [review]

File: nexus/pkg/artifacts/saver.go [hunk 1]
```diff
@@ -68,187 +53,309 @@ func (as *ArtifactSaver) createArtifact() (string, *string) {
...
+				numInProgress++
+				task := &uploader.UploadTask{
+					Path:    *entry.LocalPath,
+					Url:     *edge.Node.UploadUrl,
+					Headers: edge.Node.UploadHeaders,
+					CompletionCallback: func(task *uploader.UploadTask) {
+						taskResultsChan <- TaskResult{task, name}
+					},
+				}
+				as.UploadManager.AddTask(task)
```
Comment (raubitsj) [inline nexus/pkg/artifacts/saver.go:192]: FYI: i think we still need to add something (in a later PR) to capture uploaded files so we can pass them in filestream as completed files..  See "Uploaded" in pkg/filestream/loop_transmit.go -- and equiv logic in python internal/filestream.go

**Comment by szymon-piechowicz-wandb** [comment]
What happened here? @dmitryduev @kptkin why did you add a bunch of unrelated changes to my PR?

File: nexus/pkg/server/sender.go [hunk 2]
```diff
@@ -365,7 +365,20 @@ func (s *Sender) updateConfig(configRecord *service.ConfigRecord) {
...
+		target := s.configMap
+		for _, k := range keyList[:len(keyList)-1] {
+			val, ok := target[k].(map[string]interface{})
+			if !ok {
+				val = make(map[string]interface{})
+				target[k] = val
+			}
+			target = val
+		}
+		target[keyList[len(keyList)-1]] = value
```
Comment (szymon-piechowicz-wandb) [inline nexus/pkg/server/sender.go:381]: How did this change end up here?

File: tests/pytest_tests/system_tests/test_core/test_keras_full.py [hunk 3]
```diff
@@ -339,7 +331,6 @@ def test_keras_save_model(dummy_model, dummy_data, wandb_init):
     assert len(glob.glob(os.path.join(run.dir, "model-best.h5"))) == 1
 
 
-@pytest.mark.timeout(300)
```
Comment (szymon-piechowicz-wandb) [inline tests/pytest_tests/system_tests/test_core/test_keras_full.py:342]: How did this change end up here?

File: tests/pytest_tests/system_tests/test_core/test_public_api.py [hunk 4]
```diff
@@ -274,21 +274,20 @@ def test_run_queue(user):
 
 
 @pytest.mark.nexus_failure(feature="artifacts")
-def test_run_log_artifact(relay_server, wandb_init):
-    with relay_server():
```
Comment (szymon-piechowicz-wandb) [inline tests/pytest_tests/system_tests/test_core/test_public_api.py:278]: Why did you remove it? What server is the test talking to now?

File: wandb/apis/public.py [hunk 5]
```diff
@@ -450,7 +450,7 @@ def create_run_queue(
         access: "RunQueueAccessType",
         entity: Optional[str] = None,
         config: Optional[dict] = None,
-    ) -> None:
+    ) -> "RunQueue":
```
Comment (szymon-piechowicz-wandb) [inline wandb/apis/public.py:453]: How did this change end up here?

File: wandb/sdk/wandb_run.py [hunk 6]
```diff
@@ -583,6 +583,7 @@ def _init(
         self._config._set_settings(self._settings)
         self._backend = None
         self._internal_run_interface = None
+        # todo: perhaps this should be a property that is a noop on a finished run
```
Comment (szymon-piechowicz-wandb) [inline wandb/sdk/wandb_run.py:586]: How did this change end up here?

File: nexus/pkg/server/sender.go (hunk 2 above)
Comment (kptkin) [reply nexus/pkg/server/sender.go:381]: sorry, we were trying to release nexus-b2 and we tested it based on your PR and it looks like it got merge into your PR.

File: tests/pytest_tests/system_tests/test_core/test_keras_full.py (hunk 3 above)
Comment (kptkin) [reply tests/pytest_tests/system_tests/test_core/test_keras_full.py:342]: same as above

File: tests/pytest_tests/system_tests/test_core/test_public_api.py (hunk 4 above)
Comment (kptkin) [reply tests/pytest_tests/system_tests/test_core/test_public_api.py:278]: it will talk to the test container (see `wandb_init` fixture for details). You only need the relay_server if you want to test something related to intercepting the traffic between the client and server.

File: wandb/apis/public.py (hunk 5 above)
Comment (kptkin) [reply wandb/apis/public.py:453]: same as above

File: wandb/sdk/wandb_run.py (hunk 6 above)
Comment (kptkin) [reply wandb/sdk/wandb_run.py:586]: same as above
